src/transformers/models/bart/modeling_tf_bart.py

```diff
@@ -123,11 +125,20 @@ def __init__(self, num_embeddings: int, embedding_dim: int, **kwargs):
         self.offset = 2
         super().__init__(num_embeddings + self.offset, embedding_dim, **kwargs)
 
-    def call(self, input_shape: tf.TensorShape, past_key_values_length: int = 0):
+    def call(
+        self,
+        input_shape: Optional[tf.TensorShape] = None,
+        past_key_values_length: int = 0,
+        position_ids: Optional[tf.Tensor] = None,
+    ):
         """Input is expected to be of size [bsz x seqlen]."""
-        bsz, seq_len = input_shape[:2]
+        if position_ids is None:
+            seq_len = input_shape[1]
+            positions = tf.range(seq_len, delta=1, name="range")
+            positions += past_key_values_length
+        else:
+            positions = position_ids
 
-        positions = tf.range(past_key_values_length, seq_len + past_key_values_length, delta=1, name="range")
         return super().call(positions + self.offset)
```

Comment:
I'm not very happy with this if/else structure, but the position embedding has awkward requirements:
1. calling the layer, i.e. using its weights internally, must be done from `call`. This implies that we can't create two functions, one for each branch of the if/else case;
2. because it is a Keras layer, the first input (`input_shape`) must always be passed 🤦 

Open to suggestions here. This version is ugly, but it works.

Reply:
Looks 100% fine with me honestly don't think it's ugly

Reply:
As a nit, I'd maybe just remove the `else` statements and just always call it `position_ids`